### PR TITLE
[SYCL] Add entry table generation by wrapper for SYCL offload

### PR DESCRIPTION
**Pull request by sergey-semenov**
This patch adds "-entries" option to clang-offload-wrapper which allows
to specify a file listing all function entries that are contained within
an image. This file is generated by the sycl-post-link tool and the entry
table will then be used in SYCL runtime to select a binary based on the
requested kernel.

Signed-off-by: Sergey Semenov <sergey.semenov@intel.com>

### CODE REVIEW

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 1]
```diff
@@ -531,10 +587,20 @@ class BinaryWrapper {
...
+            std::tie(HostEntriesB, HostEntriesE);
+      else
+        std::tie(ImageEntriesB, ImageEntriesE) =
+            addSYCLOffloadEntriesToModule(Img.EntriesFile);
+
       ImagesInits.push_back(ConstantStruct::get(
           getDeviceImageTy(),
           {Fver, Fknd, Ffmt, Ftgt, Fopt, FMnf.first, FMnf.second, Fbin.first,
-           Fbin.second, EntriesB, EntriesE}));
+           Fbin.second, ImageEntriesB, ImageEntriesE}));
```
Comment (sndmitriev) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: Please remove extra braces.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 2]
```diff
@@ -554,8 +620,9 @@ class BinaryWrapper {
         ConstantInt::get(Type::getInt16Ty(C), BinDescStructVersion);
     Constant *NumImages =
         ConstantInt::get(Type::getInt16Ty(C), ImagesInits.size());
-    auto *DescInit = ConstantStruct::get(
-        getBinDescTy(), {Version, NumImages, ImagesB, EntriesB, EntriesE});
+    auto *DescInit =
+        ConstantStruct::get(getBinDescTy(), {Version, NumImages, ImagesB,
+                                             HostEntriesB, HostEntriesE});
```
Comment (sndmitriev) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: And here as well.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 3]
```diff
@@ -253,11 +263,13 @@ class BinaryWrapper {
 public:
   void addImage(const OffloadKind Kind, const llvm::StringRef File,
                 const llvm::StringRef Manif, const llvm::StringRef Tgt,
-                const BinaryImageFormat Fmt, const llvm::StringRef Opts) {
+                const BinaryImageFormat Fmt, const llvm::StringRef Opts,
+                const llvm::StringRef EntriesFile) {
```
Comment (sndmitriev) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: I guess const is not needed for StringRefs.

**Comment by sergey-semenov** [comment]
Applied comments

**Comment by sergey-semenov** [comment]
@vladimirlaz The only failure is a known problem that has been fixed, should I rebase or is this good to go?

**Comment by Fznamznon** [comment]
@sergey-semenov , please rebase your branch.

**Comment by sergey-semenov** [comment]
Rebased

**Pull request merged**